File: src/Fuaran.UI.VisualBasic.Conformance.Tests/GapClosureSlots.vb
Imports Fuaran.UI.VisualBasic

' Phase 873 — the XML-literal half of the §11 step-6 follow-up Phase 801 named. The
' gap-closure wave (861–867) added spec-record FIELDS, and the VB analyzer's pin is
' over kind NAMES, so nothing here could fail while the translator could not author
' them. These checks assert what that pin structurally cannot notice.
'
' Every check carries a negative control, because the narrowing flags are the shape
' where a vacuous pass is easiest: absent and `false` mean different things, so a
' translator that emitted `false` for an unstated column would look correct on every
' check that only asked whether the attribute reached the wire.
Module GapClosureSlots

    Sub Run(h As Harness)

        ' ── Phases 861 / 862 / 863 — the grid's behaviour declarations ──────────
        Dim grid = <DataGrid id="fleet" editable="true"
                       sort-state-key="fleet-sort"
                       default-sort-column="4" default-sort-direction="asc"
                       page-size="20" page-state-key="fleet-page"
                       edit-state-key="fleet-adjustments">
                       <Column type="text" label="Van" field="van"/>
                       <Column type="text" label="Notes" field="notes" sortable="false" editable="false"/>
                   </DataGrid>
        Dim gridJson = FuaranXml.Encode(grid)

        h.Check("861: sort-state-key reaches the wire", gridJson.Contains("""sortStateKey"":""fleet-sort"""), gridJson)
        h.Check("861: default-sort pair reaches the wire",
                gridJson.Contains("""defaultSort"":{""column"":4,""direction"":""asc""}"), gridJson)
        h.Check("862: page-size + page-state-key reach the wire",
                gridJson.Contains("""pageSize"":20") AndAlso gridJson.Contains("""pageStateKey"":""fleet-page"""), gridJson)
        h.Check("863: edit-state-key reaches the wire", gridJson.Contains("""editStateKey"":""fleet-adjustments"""), gridJson)
        h.Check("861/863: a column narrows itself out of both behaviours",
                gridJson.Contains("""editable"":false") AndAlso gridJson.Contains("""sortable"":false"), gridJson)

        ' The control: an undeclared grid declares none of it, and its columns carry
        ' NO narrowing flags — inherit is the absence, never `false`.
        Dim plain = <DataGrid id="fleet"><Column type="text" label="Van" field="van"/></DataGrid>
        Dim plainJson = FuaranXml.Encode(plain)
        h.Check("control: no behaviour keys when undeclared",
                Not plainJson.Contains("sortStateKey") AndAlso Not plainJson.Contains("pageSize") AndAlso
                Not plainJson.Contains("pageStateKey") AndAlso Not plainJson.Contains("editStateKey") AndAlso
                Not plainJson.Contains("defaultSort") AndAlso Not plainJson.Contains("sortable"), plainJson)

        ' ── Phase 801 — the static table's declared sort intent ─────────────────
        Dim table = <Table id="prices" sortable="true" default-sort-column="1" default-sort-direction="desc">
                        <Header>Item</Header>
                        <Header>Price</Header>
                        <Row><Cell>Widget</Cell><Cell>4.00</Cell></Row>
                    </Table>
        Dim tableJson = FuaranXml.Encode(table)
        h.Check("801: static sort intent reaches the wire",
                tableJson.Contains("""sortable"":true") AndAlso
                tableJson.Contains("""defaultSort"":{""column"":1,""direction"":""desc""}"), tableJson)

        Dim bareTable = <Table id="prices"><Header>Item</Header></Table>
        h.Check("801 control: no sort intent when undeclared",
                Not FuaranXml.Encode(bareTable).Contains("sortable"), FuaranXml.Encode(bareTable))

        ' ── Phase 864 — the declared field rule, all four shapes ────────────────
        Dim form = <Form id="hire">
                       <Field kind="text" id="work-email" label="Work email" required="true" rule-format="Email"/>
                       <Field kind="text" id="postcode" label="Postcode"
                           rule-pattern="[A-Z]{1,2}[0-9]" rule-message="Enter a UK postcode"/>
                       <Field kind="text" id="username" label="Username" rule-min-length="3" rule-max-length="24"/>
                       <Field kind="text" id="end-date" label="End date"
                           rule-compare-field="start-date" rule-compare-op="Gte"/>
                   </Form>
        Dim formJson = FuaranXml.Encode(form)

        h.Check("864: format rule reaches the wire", formJson.Contains("""rule"":{""format"":""email""}"), formJson)
        h.Check("864: pattern + message reach the wire",
                formJson.Contains("""pattern"":""[A-Z]{1,2}[0-9]""") AndAlso formJson.Contains("Enter a UK postcode"), formJson)
        h.Check("864: length bounds reach the wire",
                formJson.Contains("""maxLength"":24") AndAlso formJson.Contains("""minLength"":3"), formJson)
        h.Check("864: cross-field compare reaches the wire",
                formJson.Contains("""compare"":{""against"":{""$type"":""State"",""key"":""start-date""},""op"":""gte""}"), formJson)

        ' The control is load-bearing: an EMPTY rule is refused by the wire, so a
        ' translator that emitted one for every field would author documents no host
        ' accepts — and would do it silently, since nothing here reads `rule`.
        Dim ruleless = <Form id="hire"><Field kind="text" id="work-email" label="Work email" required="true"/></Form>
        h.Check("864 control: no rule when no slot is declared",
                Not FuaranXml.Encode(ruleless).Contains("""rule"""), FuaranXml.Encode(ruleless))

        ' ── Phase 867 — trend polarity, and the trend it is a statement about ───
        Dim inverted = <Metric id="wait" label="Avg wait" value="80" tone="Warning"
                           trend="-0.0734" trend-polarity="LowerIsBetter"/>
        Dim invertedJson = FuaranXml.Encode(inverted)
        h.Check("867: trend-polarity reaches the wire", invertedJson.Contains("""trendPolarity"":""LowerIsBetter"""), invertedJson)
        h.Check("867: the trend it qualifies reaches the wire too", invertedJson.Contains("""trend"":"), invertedJson)
        h.Check("867: tone survives an inverted polarity", invertedJson.Contains("""tone"":""Warning"""), invertedJson)

        ' The default is omitted on the wire, so a tile that says nothing about
        ' polarity must not start saying something.
        Dim ordinary = <Metric id="rev" label="Revenue" value="1234.5" trend="0.07"/>
        h.Check("867 control: HigherIsBetter is omitted at default",
                Not FuaranXml.Encode(ordinary).Contains("trendPolarity"), FuaranXml.Encode(ordinary))

    End Sub

End Module

File: src/Fuaran.UI.CSharp.Conformance.Tests/GapClosureSlots.cs
using System.Linq;

namespace Fuaran.UI.CSharp.Conformance.Tests;

// Phase 873 — the §11 step-6 follow-up Phase 801 named. The gap-closure wave
// (861–867) added spec-record FIELDS, and Coverage.cs reflects NodeKind CASES, so
// nothing in this suite could fail while the veneer could not author them. That is
// the whole reason these checks are written by hand: they assert what no reflection
// over the kind set can notice.
//
// The assertion is on the ENCODED WIRE SPELLING rather than on a corpus fixture,
// deliberately. The wave's grid fixtures are decoded-path shapes (`field` /
// `rowKeyField`), where a C#-authored grid carries the closure overrides the
// options records take, and the Metric fixture is bare where the smart ctor injects
// ARIA — so byte-parity against those fixtures would be a claim about the fixtures'
// authoring path, not about this veneer's. Every check below carries a negative
// control, so none can pass vacuously.
internal static class GapClosureSlots
{
    public static void Run(Harness h)
    {
        // ── Phase 861 — sortStateKey, bound defaultSort, per-column sortable ──
        var sorted = Fuaran.DataGrid(new DataGridOptions<Row>
        {
            Id = "fleet",
            ToRow = r => r.Cells,
            SortStateKey = "fleet-sort",
            DefaultSort = new DefaultSort { Column = 4, Direction = SortDirection.Asc },
            Columns = [Column.Text("Van", _ => ""), Column.Text("Notes", _ => "").Sortable(false)],
        }).Encode();

        h.Check("861: sortStateKey reaches the wire", sorted.Contains("\"sortStateKey\":\"fleet-sort\""), sorted);
        h.Check("861: bound defaultSort reaches the wire", sorted.Contains("\"defaultSort\":{\"column\":4,\"direction\":\"asc\"}"), sorted);
        h.Check("861: a column narrows itself out of sorting", sorted.Contains("\"sortable\":false"), sorted);

        // Negative control: a grid that declares none of it says none of it. The
        // narrowing flags in particular must be ABSENT rather than `false`, or every
        // unstated column would read as an explicit opt-out.
        var plain = Fuaran.DataGrid(new DataGridOptions<Row>
        {
            Id = "fleet",
            ToRow = r => r.Cells,
            Columns = [Column.Text("Van", _ => "")],
        }).Encode();

        h.Check("861 control: no sortStateKey when undeclared", !plain.Contains("sortStateKey"), plain);
        h.Check("861 control: no sortable when undeclared", !plain.Contains("sortable"), plain);
        h.Check("861 control: no defaultSort when undeclared", !plain.Contains("defaultSort"), plain);

        // ── Phase 862 — pageSize + pageStateKey ───────────────────────────────
        var paged = Fuaran.DataGrid(new DataGridOptions<Row>
        {
            Id = "members",
            ToRow = r => r.Cells,
            PageSize = 20,
            PageStateKey = "members-page",
            Columns = [Column.Text("Month", _ => "")],
        }).Encode();

        h.Check("862: pageSize reaches the wire", paged.Contains("\"pageSize\":20"), paged);
        h.Check("862: pageStateKey reaches the wire", paged.Contains("\"pageStateKey\":\"members-page\""), paged);
        h.Check("862 control: no paging when undeclared", !plain.Contains("pageSize") && !plain.Contains("pageStateKey"), plain);

        // ── Phase 863 — editStateKey + per-column editable ────────────────────
        var editable = Fuaran.DataGrid(new DataGridOptions<Row>
        {
            Id = "stock",
            ToRow = r => r.Cells,
            Editable = true,
            EditStateKey = "stock-adjustments",
            Columns = [Column.Text("Month", _ => ""), Column.Text("Note", _ => "").Editable(false)],
        }).Encode();

        h.Check("863: editStateKey reaches the wire", editable.Contains("\"editStateKey\":\"stock-adjustments\""), editable);
        h.Check("863: a column narrows itself read-only", editable.Contains("\"editable\":false"), editable);
        h.Check("863 control: no editStateKey when undeclared", !plain.Contains("editStateKey"), plain);

        // ── Phase 801 — the static table's declared sort intent ───────────────
        var table = Fuaran.Table(new TableOptions
        {
            Id = "prices",
            Headers = [(Text)"Item", (Text)"Price"],
            Rows = [[(Text)"Widget", (Text)"4.00"]],
            Sortable = true,
            DefaultSort = new DefaultSort { Column = 1, Direction = SortDirection.Desc },
        }).Encode();

        h.Check("801: static sortable reaches the wire", table.Contains("\"sortable\":true"), table);
        h.Check("801: static defaultSort reaches the wire", table.Contains("\"defaultSort\":{\"column\":1,\"direction\":\"desc\"}"), table);

        var bareTable = Fuaran.Table(new TableOptions { Id = "prices", Headers = [(Text)"Item"] }).Encode();
        h.Check("801 control: no sort intent when undeclared", !bareTable.Contains("sortable") && !bareTable.Contains("defaultSort"), bareTable);

        // ── Phase 864 — FormField.rule, all four shapes plus the message ──────
        var form = Fuaran.Form(new FormOptions
        {
            Id = "hire",
            Fields =
            [
                FormField.Text("work-email", "Work email", required: true, rule: new FieldRule { Format = TextFormat.Email }),
                FormField.Text("postcode", "Postcode", rule: new FieldRule
                {
                    Pattern = "[A-Z]{1,2}[0-9][A-Z0-9]?",
                    Message = "Enter a UK postcode",
                }),
                FormField.Text("username", "Username", rule: new FieldRule { MinLength = 3, MaxLength = 24 }),
                FormField.Text("end-date", "End date", rule: new FieldRule
                {
                    Compare = CompareRule.AgainstField("start-date", CompareOp.Gte),
                }),
            ],
        }).Encode();

        h.Check("864: format rule reaches the wire", form.Contains("\"rule\":{\"format\":\"email\"}"), form);
        h.Check("864: pattern + message reach the wire", form.Contains("\"pattern\":\"[A-Z]{1,2}[0-9][A-Z0-9]?\"") && form.Contains("Enter a UK postcode"), form);
        h.Check("864: length bounds reach the wire", form.Contains("\"maxLength\":24") && form.Contains("\"minLength\":3"), form);
        h.Check(
            "864: cross-field compare reaches the wire",
            form.Contains("\"compare\":{\"against\":{\"$type\":\"State\",\"key\":\"start-date\"},\"op\":\"gte\"}"),
            form);

        var ruleless = Fuaran.Form(new FormOptions
        {
            Id = "hire",
            Fields = [FormField.Text("work-email", "Work email", required: true)],
        }).Encode();
        h.Check("864 control: no rule when undeclared", !ruleless.Contains("\"rule\""), ruleless);

        // ── Phase 867 — trendPolarity ─────────────────────────────────────────
        var inverted = Fuaran.Metric(new MetricOptions
        {
            Id = "wait",
            Label = "Avg wait",
            Value = 80.0,
            Trend = -0.0734,
            TrendPolarity = TrendPolarity.LowerIsBetter,
            Tone = Tone.Warning,
        }).Encode();

        h.Check("867: trendPolarity reaches the wire", inverted.Contains("\"trendPolarity\":\"LowerIsBetter\""), inverted);

        // Negative control, and it is the one that matters most here: the DEFAULT is
        // omitted, so a veneer that hard-coded the enum would look identical on a
        // default tile and differ only on the inverted one.
        var ordinary = Fuaran.Metric(new MetricOptions { Id = "rev", Label = "Revenue", Value = 1.0, Trend = 0.07 }).Encode();
        h.Check("867 control: HigherIsBetter is omitted at default", !ordinary.Contains("trendPolarity"), ordinary);

        // And polarity must not disturb `tone` — one slot could never have said both.
        h.Check("867: tone survives an inverted polarity", inverted.Contains("\"tone\":\"Warning\""), inverted);
    }

    /// <summary>A minimal row type for the grid checks — the veneer's REQUIRED
    /// <c>ToRow</c> projection has to be satisfied, and the checks above are about
    /// the behaviour slots rather than the cells.</summary>
    private sealed record Row
    {
        public System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object>> Cells =>
            Enumerable.Empty<System.Collections.Generic.KeyValuePair<string, object>>();
    }
}
